app/src/main/java/com/bignerdranch/android/photogallery/QueryPreferences.kt

```diff
@@ -0,0 +1,22 @@
+package com.bignerdranch.android.photogallery
+
+import android.content.Context
+import android.preference.PreferenceManager
+import androidx.core.content.edit
+
+object QueryPreferences {
+
+    fun getStoredQuery(context: Context): String {
+        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
+        return prefs.getString(PREF_SEARCH_QUERY, "")!!
```

Comment:
this in general is error-prone. The double exclamation mark (`!!`) means you are by all means expecting a nullable value to be non-null. If for some reason it is `null`, and the return type of `getStoredQuery` expects a non-null value, using `!!` will cause a runtime error. So, in here you're saying something like _hey, get me this string and I'm 100% sure it is there, where you're getting it from_. It's a good practice not to force it like that using `!!`, since it can easily lead to runtime exceptions

Reply:
_the return type for SharedPreferences.getString(...) is defined as a nullable String because the compiler cannot guarantee that the value associated with PREF_SEARCH_QUERY exists and that it is not null. But you know that you never store a null value for PREF_SEARCH_QUERY - and you supply an empty String as the default value in cases where setStoredQuery(context: Cintext, query: String) has not yet been called. So it is safe to use the non-null assertion operator here without a try/catch block around it._ That's what book says. Wouldn't be better to not use `!!` in this case or in general?

Reply:
In general, its better not to use the `!!`. If thats what the docs say then sure